zerver/tests/test_default_avatars.py

```diff
@@ -0,0 +1,390 @@
+        # Version should increment for cache busting
+        self.assertGreater(hamlet.avatar_version, initial_version)
+
+    def test_realm_creation_default_provider(self) -> None:
+        """Test that new realms default to Jdenticon."""
+        realm = do_create_realm("test-default", "Test Default Realm")
+        # Should default to Jdenticon (1)
+        self.assertEqual(realm.default_avatar_provider, 1)
+
+    def test_events_sent_on_provider_change(self) -> None:
+        """Test that proper events are sent when provider changes."""
+        self.login("iago")
+        hamlet = self.example_user("hamlet")
+        hamlet.avatar_source = "J"
+        hamlet.save(update_fields=["avatar_source"])
+
+        # Change provider - events will be sent
+        result = self.client_patch(
+            "/json/realm",
+            {"default_avatar_provider": 3},
+        )
+        self.assert_json_success(result)
+
+        # Verify realm setting changed
+        realm = get_realm("zulip")
+        self.assertEqual(realm.default_avatar_provider, 3)
+
+        # Verify user was updated
+        hamlet = get_user_profile_by_id(hamlet.id)
+        self.assertEqual(hamlet.avatar_source, "S")
+
+    @override_settings(AVATAR_CHANGES_DISABLED=True)
+    def test_avatar_changes_disabled_setting(self) -> None:
+        """Test behavior when avatar changes are disabled."""
+        self.login("iago")
+
+        # Admin should still be able to change default provider
+        result = self.client_patch(
+            "/json/realm",
+            {"default_avatar_provider": 2},
+        )
+        # This should succeed since it's a realm-wide setting, not individual avatar upload
+        self.assert_json_success(result)
+
+    def test_deactivated_users_not_updated(self) -> None:
+        """Test that deactivated users don't get avatar updates."""
+        self.login("iago")
+        hamlet = self.example_user("hamlet")
+        hamlet.avatar_source = "J"
+        hamlet.save(update_fields=["avatar_source"])
+        change_user_is_active(hamlet, False)
+
+        initial_version = hamlet.avatar_version
+
+        # Change provider
+        result = self.client_patch(
+            "/json/realm",
+            {"default_avatar_provider": 3},
+        )
+        self.assert_json_success(result)
+
+        hamlet = get_user_profile_by_id(hamlet.id)
+        # Deactivated user should not be updated
+        self.assertEqual(hamlet.avatar_source, "J")
+        self.assertEqual(hamlet.avatar_version, initial_version)
+
+    def test_bot_users_updated_with_provider_change(self) -> None:
+        """Test that bot users are also updated when provider changes."""
+        self.login("iago")
+        bot = self.create_test_bot("test-bot", self.example_user("hamlet"))
+        bot.avatar_source = "J"
+        bot.save(update_fields=["avatar_source"])
+
+        # Re-login as iago (create_test_bot may have changed the session)
+        self.login("iago")
```

Comment:
This is a bad testing pattern, use `self.api_patch` instead of changing who is logged in constantly.